docs/rfcs/20220131_raft_snapshots.md

```diff
@@ -0,0 +1,149 @@
+
+A leader decides to do compaction at offset X:
+ 
+  1. saves current snapshot
+  2. waits until it's safe to do compaction:
+    a. segment containing X is closed
+    b. tx front is passed the end of the segment
+  3. replicates snapshot
+ 
+A leader or a follower sees the snapshot via `state_machine::apply` (it happens only when the batch is replicated to the majority and can't be lost):
+
+  - perform the compaction when current node is the leader and it sent the snapshot (within current term)
+  - when the snapshot isn't applied:
+    1. rewind log at offset X
+    2. apply the snapshot
+  - when the snapshot is applied:
+    - perform the compaction
+
+There is a problem with this approach: a follower doesn't know if the batches it replays contain holes so its current state may be incorrect until it sees the snapshot batch. This problem becomes actual when Redpanda supports reads from replicas. There are several ways to deal with it.
+
+#### Blocking replica reads
+
+A follower is allowed to serve reads only up to the last observed snapshot. Even when the prior batches contain holes they become visible when the snapshot is applied and corrected.
+
+#### Modify the `append_entries_request`
+
+We may modify the append_enties_request to include an offset of the last compaction. When a follower sees that last compaction less than current offset it knows that the latest batches don't have holes and it's safe to serve the requests.
+
+When we modify `append_entries_request` or introduce and switch to the new API we need to thing of the update story. In order to avoid shutting down a cluster for the update we need to put the change behind a feature flag, update the brokers one by one and only then to active the feature.
+
+## Approach #2: Ship the snapshot (cached state) via the new API
+
+An alternative idea is to use a side channel instead of the main replication path to ship the snapshot. With this approach we don't shift the offset and it eventually opens a way to get rid of the offset translation layer.
+
+### Overview
+
+When a leader decides to compact the log at offset X it:
+
+  1. makes a snapshot and sends it to a majority of nodes via new API
+  2. waits until it's safe to do compaction:
+    a. segment containing X is closed
+    b. tx front is passed the end of the segment
+  3. starts include the new offset X to the `append_entries_request`
+  4. performs the compaction
+ 
+When a follower observes a compacted offset past its snapshot via `append_entries_request` it:
+
+  1. queries all nodes and finds a node which has the snapshot (there is at least a majority of such nodes so availability stays the same as with Raft)
+  2. downloads the snapshot
+  3. applies the snapshot
+  4. rewinds the snapshot to the snapshot's offset
+  5. acknowledges the `append_entries_request`
+
+It's safe to serve read from a replica bounded by the last snapshot's offset or by current offset if the compacted offset is equal to the last snapshot.
```

Comment:
is "serve read" a typo?

Reply:
yes, I think it should be `to serve read requests` .